[Graph] Add `acos()` Max operation

msaelices opened the pull request:
This adds the arccosine (inverse cosine) operation to MAX graph ops. The implementation uses a compositional approach with existing graph operations, following the Remez polynomial approximation from math.mojo.

msaelices committed:
[Graph] Add acos() operation

This adds the arccosine (inverse cosine) operation to MAX graph ops.
The implementation uses a compositional approach with existing graph
operations, following the Remez polynomial approximation from math.mojo.

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Move the test to the right place

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[Graph] Improve test_acos.py with actual execution tests

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

abduld commented on max/graph/ops/acos.py:
```diff
@@ -0,0 +1,126 @@
+# ===----------------------------------------------------------------------=== #
```
this is not super efficient and duplicates the mojo implementation. It would be better to just introduce a custom kernel which is a wrapper around foreach and math.acos

msaelices replied:
I'm not sure if I did correctly but it's done in these commits:
https://github.com/modular/modular/pull/5418/commits/624035ab58f637d9582b53cbe190558d9adc2ffd
https://github.com/modular/modular/pull/5418/commits/ed2517706b099d0d6f7adf42b26cc78521755c0d

abduld replied:
Thanks, this is almost how to do it. Can you move the acos function to the elementwise file

msaelices committed:
[Kernels] Implement acos operation using custom kernel

This replaces the compositional graph-based implementation of acos
with an efficient custom kernel that directly calls math.acos() from
the Mojo standard library using the foreach pattern.

msaelices committed:
[Kernels] Fix a circular import error in max acos()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

abduld commented on max/graph/ops/acos.py:
```diff
@@ -0,0 +1,57 @@
+    from . import dtype_promotion  # Avoid circular import
+
+    x = dtype_promotion._restrict_to_strong_dtypes(x)
+    return Graph.current._add_op(mo.mo_acos, x._mlir_value.type, x)[0].tensor
```
there is no mo op called acos, so you want to add this as a custom op via `ops.custom`. It should look something like

```
return ops.custom(
     name="mo.acos",
     device = device,
     values = [x],
     out_types = 
                TensorType(
                    dtype=x.dtype,
                    shape=x.tensor.shape,
                    device=DeviceRef.from_device(device),
                )
 )[0].tensor
```

msaelices committed:
[Graph] Change the acos() implementation to add a custom ops

Thanks to @abduld for the tip

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[Graph] Fix circular import error

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[Graph] Fix imports in acos() Max op

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'main' into max-acos-op
